tools/16ton/parport/pport-0.6.9/man/pport.1
.\" Copyright 2004 Corcalciuc V. Horia (rain@sevenrains.ro)
.\" May be distributed under the GNU General Public License
.TH PPORT 1 "20 January 2004"
.SH NAME
pport \- manipulate computer port pins
.SH SYNOPSIS
.nf
.BR "pport <" flippers "> [" modifiers "]" 
.BR "pport <" extensions ">"
.BR "pport <" information ">"
.fi
.SH DESCRIPTION
The program 
.B pport 
is a simple yet handy program and its documented counterpart for
accessing the output pins of multiple ports. Using this bundle, one can
successfully control any household appliance or electronic device with
minimal hassle and practically no changes. 
.B pport
will alter the state (toggle or set) of the data pins specified by the
user, of the parallel port or other supported ports using 
.B flippers 
to toggle or set, 
.B modifiers 
to alter the methods and provides 
.B information 
all availible from the contextual 
.B \--help 
or 
.B \-h 
options directly from the command line.

Conceived to change between the low or high state of the data pins 
of the parallel port, development extended its use to other
common PC ports:

The i386 parallel port is a 25 pin D-shaped female connector 
in the back of the computer, conventionally used to connect a printer.
Not all 25 pins are always needed. Thus it can alter the state of eight of 
those pins called data pins: 2, 3 4, 5, 6, 7 and 8. The rest are simple 
grounding pins as seen in the illustration below.
Here is a scheme of the parallel port:

.nf
                                 *-data pins D0 to D7

                        |D7 |D6 |D5 |D4 |D3 |D2 |D1 |D0    <---+          
   _____________________|___|___|___|___|___|___|___|______    | count     
  *                     |   |   |   |   |   |   |   |       *  |     
  *    13  12  11  10   9   8   7   6   5   4   3   2  1    *  |     
  *                                                         *        
   *     25  24  23  22  21  20  19  18  17  16  15  14    *         
    *____|___|___|___|___|___|___|___|____|_______________*          
         |   |   |   |   |   |   |   |    |                          
          +---+---+---+---+---+---+---+----+-----+                    
                           *                     |
                           |                     |
                 groudning pins 17 to 25        ---                   
                                               ////                   

.fi

The data pins are D0 through D7. Pins 17 through 25 are the grounding pins.

The program also has support for the serial rs232 port. Although the IO
is different from that of the parallel port, the serial connector rs232 
can prove useful in many situations since it offers a serious amount 
of power (a voltage of +-12V or in some cases +-5V  and short circuit 
current flow of 7-10mA) depending on the circuit types used in the PC 
circuitry. This is all at the cost of signal lines. The parallel port has 
eight data pins as we have seen above whereas the serial port has only 
three: TD(Transmit Data), DTS(Data Terminal Ready), RTS(Request to Send).
However, only the last two are usable since TD is logic true(-12V) when 
no data is sent:

.nf                           

   9 pin rs232           +-DTR       In order to avoid
         _______________|_______    confusion with the
        *                |       *  other parallel port        
        *    1   2   3   4   5   *  we only mention that    
         *     6   7   8   9    *   the 25-pin version of                 
          *________|___________*    the serial port has:    
                   |                    DTR = pin 20
                   +_RTS_               RTS = pin 7

.fi

Based on the equivalence of both parallel port versions, 
.B pport
accepts both standards transparently, either 9 pin or 25 pin serial rs232 
connector.

.SH OPTIONS
.TP
.B \-p, \--port <serial>
With this option one can select an alternative port.
.TP
.B \-s, \--set <pin1,pin2...|all|none|even|odd|random|dtr,rts>
Causes
.B pport
to set one, some, or all pins to high. Or to set all low with "none". dtr and
rts are available only in combination with the serial port. One can also use
all and none with the serial port option in order to set both RTS and DTR
lines to high respectively low.
.TP
.B \-t, \--toggle  <pin1,pin2...|all|even|odd|random>
Causes
.B pport
to toggle the state of the specified pins.
.TP
.B \-w, \--wait  <Nd|Nh|Nm|Ns>
Where n is a number, causes
.B pport
to wait a specific time before altering the state of the specified pins, 
where n is a number, and d stands for days, h for hours, m for minutes and s
for seconds.
.TP
.B \-r, \--restore
Causes
.B pport
to restore the previous state of the pins stored in the data file.
.TP
.B \--status
Makes
.B pport
give a status report on the parallel port.
.TP
.B \--debug
Will turn on debugging for the following command.
.TP
.B \--raw
Will send raw data through the parallel port.
.SH EXAMPLES
Here are a few examples for using this program:
.TP
pport -s all -w 10m
Will set all the pins to high after a 10 minutes delay.
.TP
pport -s random -w 5s
Will set some random pins to high after a 5 seconds delay.
.TP
pport -r
Will restore the previous pin states. This can be useful if you want
to include pport in your startup scripts so it can restore your device
status on system startup.
.TP
pport -t 2,4,7 -w 10s
Will toggle pins 2, 4 and 7 into their complementary state after 10
seconds.
.TP
pport -p serial -s rts
Will set the rts line of the rs232 port to the high state.
.TP
pport -p serial -s rts,dtr
Will set both lines of the rs232 port to the high state.
.TP
pport -p serial -s none
Will set both lines of the rs232 port to the low state.
.TP
pport --raw "Hello world"
Will send data through the parallel port.
.SH NOTES
1. The
.B pport
utility, must be able to access the ports in the first place.
Failure to do so will result in an error. Thus
.B pport
must have superuser priviledges at runtime.
.SH HISTORY
This package is the uttermost practical embodiment of the Coffee-HOWTO
originally written by Fotis Georgatos, <gef@hellug.gr> and Annie Pinder, <
ajp@leguin.org.uk>. It tries to provide a standard for controlling
computer ports quickly aiming at the same time to be simple. The "custom" howto
provided, notably FRANKY-HOWTO, contains various hardware applications
that can be used. It provides an updated set of ideas that could maybe
realize the functionality you seek.
.SH AUTHORS
.nf
Written by Corcalciuc V. Horia (rain@sevenrains.ro)
.fi


debian/changelog
linux-4.14.13-16ton-gpd-win-gaming-conf (4.14.13-16ton-gpd-win-gaming-conf-1) xenial; urgency=low

  * Custom built Linux kernel.

 -- Anonymous <root@maddocks-Q552UB>  Thu, 11 Jan 2018 12:07:50 -0500


tools/16ton/parport/pport-0.6.9/src/pport.c
/*
        pport: Small program to interface your computer with the
               parallel port. Inspired from the Coffee-Machine HOWTO
        
        Copyright (C) 2003  Corcalciuc V. Horia

        This library is free software; you can redistribute it and/or
        modify it under the terms of the GNU Lesser General Public
        License as published by the Free Software Foundation; either
        version 2.1 of the License, or (at your option) any later version.
                
        This library is distributed in the hope that it will be useful,
        but WITHOUT ANY WARRANTY; without even the implied warranty of
        MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
        Lesser General Public License for more details.
                                
        You should have received a copy of the GNU Lesser General Public
        License along with this library; if not, write to the Free Software
        Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307 USA 



     

*/

/* System configuration */

#include "config.h"

/*	Basic includes.	*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <popt.h>
#include <ctype.h>
#include <time.h>
#include <errno.h>
#include <unistd.h>

/*	Our includes.	*/

#include "defines.h"

int
main (const int argc, const char **argv)
{

  enum bool
  {
    false = 0,
    true
  } verboose = false;

  int itra=0;
  int result=0;
  int pid=0, status=0;
  char *optBuf=NULL;
  const char *_port_;
  const char *optData=NULL;
  const char *delay=NULL;

  poptContext optCon;

  const struct poptOption optionT[] = {

    {"\b\b  \nFlippers:", 0, 0, NULL, 0, NULL},
    {"set", 's', POPT_ARG_STRING, &optData, 's',
     "set pins", "[P|S]"},
    {"toggle", 't', POPT_ARG_STRING, &optData, 't',
     "toggle pins", "[P]"},
    {"restore", 'r', 0, NULL, 'r',
     "restore status", NULL},
    {"\b\b  \nModifiers:", 0, 0, NULL, 0, NULL},
    {"wait", 'w', POPT_ARG_STRING, &delay, 'w',
     "delay", "[ALL]"},
    {"\b\b  \nInformation:", 0, 0, NULL, 0, NULL},
    {"status", 0, 0, NULL, 'm',
     "print status", NULL},
    {"debug", 0, 0, NULL, 'd',
     "debugging", NULL},
    {"\b\b  \nExtensions:", 0, 0, NULL, 0, NULL},
    {"raw", 0, POPT_ARG_STRING, &optData, 'k',
      "send raw data", "[STRING]"},
    {"\b\b  \nHelp:", 0, 0, NULL, 0, NULL},
    {"help", 'h', 0, NULL, 'h',
     "display quick help", NULL},
    {"man", 0, 0, NULL, 'i',
     "display manual page", NULL},
    {NULL, 0, 0, NULL, 0}

  };

  _port_ = (char *) malloc (10 * sizeof (char));
  optData = (char *) malloc (strlen (ALL) * sizeof (char));
  delay = (char *) malloc (strlen (ALL) * sizeof (char));

  optBuf = (char *) malloc (strlen (*argv) * sizeof (char));


  optCon = poptGetContext (NULL, argc, argv, optionT, 0);

  if (argc < 2)
    {

      poptPrintUsage (optCon, stderr, 0);
      exit (1);

    }

  while ((result = poptGetNextOpt (optCon)) >= 0)
    switch (result)
      {
      case 'h':
        optBuf[itra++] = 'h';
        break;
      case 'i':
	    optBuf[itra++] = 'i';
	    break;
      case 'y':
        optBuf[itra++] = 'y';
        break;
      case 's':
        optBuf[itra++] = 's';
        break;
      case 't':
        optBuf[itra++] = 't';
        break;
      case 'o':
        optBuf[itra++] = 'o';
        break;
      case 'w':
        optBuf[itra++] = 'w';
        break;
      case 'r':
        optBuf[itra++] = 'r';
        break;
      case 'm':
        optBuf[itra++] = 'm';
        break;
      case 'd':
        optBuf[itra++] = 'd';
        break;
      case 'k':
        optBuf[itra++] = 'k';
        break;
	}

  if (result < -1)
    {
      fprintf (stderr, "%s: %s\n",
	       poptBadOption (optCon, POPT_BADOPTION_NOALIAS),
	       poptStrerror (result));
      return 1;
    }
  if(strchr (optBuf, 'i') != NULL)
    {	
  		if((pid=vfork())==0)
				execlp("man", "-a", PACKAGE_NAME, NULL);
			else if(pid<0) fprintf(stdout, "%s", strerror(errno));
			while(wait(&status)<0);
			exit(0);
	} 
  if (strchr (optBuf, 'h') != NULL)
    {
      poptPrintHelp (optCon, stderr, 0);
      fprintf (stderr, "\n");
      fprintf (stderr, "\tP\tvalid: all, none, even, odd,\n\t\trandom or 2,3,4, etc...\n");
      fprintf (stderr, "\tS\tfor serial port: dtr, rts\n");
      fprintf (stderr, "\tALL\tNd N days, Nh N hours,\n\t\tNm N minutes, Ns N seconds.\n");
      exit (0);
    }   

  if (strchr (optBuf, 'd') != NULL)
    {
      verboose = true;
    }
  if (strchr (optBuf, 'w') != NULL)
    {
      result = _getdelay (delay);
    }
  if (strchr (optBuf, 's') != NULL)
    {
      if (verboose)
	    fprintf (stdout, "Set mechanism: %s\n", optData);
      if (strstr(optData, "dst")!=NULL || strstr(optData, "rts")!=NULL)
	    result = _set (optData, SBASE);
      else
	    result = _set (optData);
    }
  else if (strchr (optBuf, 'r') != NULL)
    {
      _restore ();
    }

  else if (strchr (optBuf, 't') != NULL)
    {

      if (verboose == true)
	fprintf (stdout, "Toggle pins: %s\n", optData);
      result = _toggle (optData);

    }
  else if (strchr (optBuf, 'm') != NULL)
    {

      result = _status ();

    }
  else if (strchr (optBuf, 'k') != NULL)
    {
      result = raw_snd (optData);
    }
  else
    poptPrintHelp (optCon, stderr, 0);

  if (optCon)
    poptFreeContext (optCon);

  return 0;

}

/*	EOF	*/
